### ISSUE TITLE
[Performance] Improve Android Cold Boot TTI (Time To Interactive)

### ISSUE THREAD

**Issue by marcaaron** [Expensify/App#4656]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## What performance issue do we need to solve?

- App launch and time to interactive
- Issue feels likely to be in the JS layer

## What is the impact of this on end-users?

The app takes a long time to become interactive on Android. 
We know that the number of reports a user has can impact the severity.

## List any benchmarks that show the severity of the issue

We can test this in an Android release build by using the `CAPTURE_METRICS=true` option in `.env` file and looking at the `timeToInteractive` metric after booting up the app (must be previously signed in). 

**Follow Instructions here:**
https://github.com/Expensify/App/blob/main/PERFORMANCE.md#performance-metrics-opt-in-on-local-release-builds

**Android app startup JS sampling profile to inspect in Chrome Dev Tools (Taken from a Galaxy S8):**

[sampling-profiler-trace4670664525343774149-converted.json.zip](https://github.com/Expensify/App/files/6985380/sampling-profiler-trace4670664525343774149-converted.json.zip)

There's also a lot of context in [this issue](https://github.com/Expensify/App/issues/4027).

## Proposed solution (if any)
N/A

## List any benchmarks after implementing the changes to show impacts of the proposed solution (if any)
N/A

## Platform:
Where is this issue occurring?

- [ ] Web
- [ ] iOS
- [x] Android
- [ ] Desktop App
- [ ] Mobile Web

**Version Number:**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
**Expensify/Expensify Issue URL:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @bfitzexpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Luke9389 (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by bfitzexpensify** [comment]
Upwork [posting](https://www.upwork.com/jobs/~01c0795b8527047766)

**Comment by kidroca** [comment]
I've managed to hook the flipper plugin to print this information instead of the Alert https://github.com/oblador/react-native-performance/blob/master/packages/flipper-plugin-performance/README.md
- the newer plugin is flipper-plugin-performance (not flipper-plugin-react-native-performance (deprecated))

![Screenshot 2021-08-16 at 22 33 15](https://user-images.githubusercontent.com/12156624/129620487-474f9e5d-ebbf-4ffd-8130-21e1ce3d0a39.png)

You can mark more events and they would appear on the timeline as well e.g. report selected -> report rendered 

This also allows exporting and importing files to Flipper 
[FlipperExport.flipper.zip](https://github.com/Expensify/App/files/6995055/FlipperExport.flipper.zip)

**Comment by marcaaron** [comment]
@kidroca is there some way to use Flipper on a release build? I've found that these metrics are much different on dev builds which is why the alert method is used here.

**Comment by kidroca** [comment]
Any updates made on the ReportScreen or it's children doesn't seem to affect this metric as the end mark for TTI happens before the ReportScreen mounts (or shortly after it starts mounting)

> @kidroca is there some way to use Flipper on a release build? I've found that these metrics are much different on dev builds which is why the alert method is used here.

I'm not sure how, but it's probably possible, I remember I've done iOS prod builds where the dev menu is enabled
For Android you can switch between PROD and DEV js bundle from the menu and it makes a big difference 
![image](https://user-images.githubusercontent.com/12156624/129633418-8946cfef-0606-4f41-822a-a246ddee6498.png)

I've added marks for ReportActionsView.
Marks: LHN click -> RAV mount -> RAV layout, here's how they look using PROD js 
![image](https://user-images.githubusercontent.com/12156624/129632312-f0096535-77a1-4569-84c8-2409db60e3e8.png)
- the last 2 switches taking less time - these are switching back to one of the previous two reports

And here's the same for a DEV build 
![image](https://user-images.githubusercontent.com/12156624/129631079-fafea1ce-4292-4893-ae5c-04b3ed55a1fa.png)
 - here I can't capture all interactions on a single screen, but they are similar to the one on the left 2.1sec, the last 2 are again switch backs

**Comment by marcaaron** [comment]
>For Android you can switch between PROD and DEV js bundle from the menu and it makes a big difference

It's good to have the bundle optimized for release, but it takes time to download via the metro server. This is important when we are timing a cold start from app icon tap to interactive so release builds have been most reliable for me.

That said, I don't mind if we enable this flipper plugin or add additional metrics if we think that will be helpful.

**Comment by marcaaron** [comment]
@bfitzexpensify just a heads up that we can accept multiple proposals for this one. I don't think there will be a single solution to this one so mainly looking to accept any proposal that can improve this metric (which could mean the issue has multiple assignees).

**Comment by kidroca** [comment]
Captured some more insights as to where we're spending time during init

This is a PROD js build (from the dev menu - not a release build)
![image](https://user-images.githubusercontent.com/12156624/129711227-e3c66d08-d48e-4d36-bdad-1c2b4514f651.png)

I'm 99% sure the red 🔴 needles are showing initial script parse and execution. Though I couldn't find documentation confirming this.
Everything before the 1st 🔴 needle is probably something we can ignore as we can't really affect it by changing the JS. 
That's also mentioned inside the plugin code:
> // The bundle download trace can be very long but has no real impact on
// perf so we try to alter the marks to act as if it wasn't there

We're seeing so far before the 1st needle because of the measures we've made, otherwise the timeline would start from the first needle

Slightly before the 2nd 🔴 needle are the first Onyx calls, they are part of the initial script execution  - `Onyx.init` is not called inside a `componentDidMount` but a as a top level call, so it's considered a part of the initial script execution (Maybe we can try to move it inside a component lifecycle call to see if this gains anything)

700ms past the 2nd 🔴 needle is the TTI event
If we measure from First onyx call (at 7.700) to TTI (at 9.100) it's about 1.4sec 

It's really easy to update the Onyx metrics to be captured using `react-native-performance`, visualized along some other interactions it provides an interesting insight - should I open a PR about this? 
I've also enabled the networking metrics 

Here's the export from flipper, you'll need to have the `Performance` plugin installed to be able to see it 
[FlipperExport.zip](https://github.com/Expensify/App/files/6999487/FlipperExport.zip)

**Comment by kidroca** [comment]
From Marc's screen in the description the `runJsBundle` time is a lot less 353ms, compared to mine 4200ms 
This can vary from device to device, but maybe the huge diff is because there are still some debugging tools running 

IMO we can use the debug build to test theories and then use the release build (before/after) to prove them

**Comment by marcaaron** [comment]
I like the idea to capture Onyx metrics with the Flipper plugin. I haven't relied on them very much tbh so maybe having them enabled when `CAPTURE_METRICS` is `true` will make them more discoverable for anyone investigating performance issues.

>From Marc's screen in the description the runJsBundle time is a lot less 353ms, compared to mine 4200ms

I think it's because the JS source is precompiled into bytecode for the release build when using Hermes engine? Which perhaps does not happen when we build the debug app. But unsure tbh.

>IMO we can use the debug build to test theories and then use the release build (before/after) to prove them

I agree with this, but at the end of the day the release build never lies 😄

**Comment by bfitzexpensify** [comment]
>@bfitzexpensify @Luke9389 just a heads up that we can accept multiple proposals for this one. I don't think there will be a single solution to this one so mainly looking to accept any proposal that can improve this metric (which could mean the issue has multiple assignees).

Sure thing — just let me know who we want to hire and I'll keep the posting open (same to you @Luke9389)

**Comment by kidroca** [comment]
> I think it's because the JS source is precompiled into bytecode for the release build when using Hermes engine?

Correct. 

> To see the benefits of Hermes, try making a release build/deployment of your app
> This will compile JavaScript to bytecode during build time which will improve your app's startup speed
https://reactnative.dev/docs/next/hermes#confirming-hermes-is-in-use

0.35 vs 4.2 is a big improvement.

I'll check if there's some way to use Flipper to see the performance results of a release build 

An alternative is that a debug build can be configured to precompile JS the same way a release does. Saw that in one of the perf capturing tutorials, there's a gradle config about it. This would keep Flipper available, and skip having to create certificates to sign release builds

I wonder why on iOS a dev build still seems significantly faster in every aspect
We try to pump everything possible out of Android doing a release build and still can't catch up to iOS on a dev build

**Comment by kidroca** [comment]
Application bundling can be updated to use RAM bundles to improve TTI in general
This means extracting non essential parts of the application to separate chunks that are lazy loaded on demand

Related Article: https://reactnative.dev/docs/ram-bundles-inline-requires

We load all the screens and navigators even though only 1 is used during init, and some might not be used at all in a single session
Using RAM bundles and inline requires we can optimize the app to load such components the first time we navigate to a given route 

My idea is to 1) make a crude check on potential improvements by removing all navigators but the MainDrawerNavigator 
and if it indeed is beneficial 2) Make a `LazyNavigator` component that would encapsulate the inline require logic and would be applied over existing navigators

During 1) We can also capture memory usage with/without navigators as besides TTI this should potentially improve memory by offloading unused screens

**Comment by marcaaron** [comment]
I did some digging on RAM bundles at one point and the discovered they aren't compatible when using Hermes. Something something `mmap`...

![Screen Shot 2021-08-18 at 6 53 47 AM](https://user-images.githubusercontent.com/32969087/129939862-0e04c99e-3735-4308-9d72-aa50b741edd9.png)

Inline requires should still work with Hermes (I know I read this somewhere but didn't think of any great ways to implement) so that's something we could try. I'd definitely be curious to know if there are some large libraries or things doing a lot of work that could be lazy loaded or deferred.

As for `react-navigation`, I read something somewhere (SO or Reddit hearsay I think) that Android is "not as efficient" as iOS at handling the huge tree of components that `react-navigation` necessarily creates. I followed some suggestions to use the [native stack navigator](https://reactnavigation.org/docs/native-stack-navigator/). Which seemed to help marginally, but also limited the flexibility of screen animations. Cutting `react-navigation` out of the picture entirely (and just rendering the sidebar on init without anything else) did yield about a ~400 ms reduction in TTI.

**Comment by kidroca** [comment]
I've submit a PR here to review the Onyx metrics inside Flipper: https://github.com/Expensify/react-native-onyx/pull/101
I'll submit another one that adds the Flipper Performance plugin and captures the ReportActionsView layout in Expensify/App

**Comment by kidroca** [comment]
It's **not** possible to use Flipper with a release build (or at least without tweaking too many things) 
> "Question: Can I use flipper on a release build on Android/iOS?" - https://github.com/facebook/flipper/issues/1488

It's relatively easy to include a compiled bundle (bytecode) in the dev build (it's still not the same as running a release build, but comes close)
![image](https://user-images.githubusercontent.com/12156624/130079321-314e0311-e004-4388-a888-8739af9b2646.png)
![image](https://user-images.githubusercontent.com/12156624/130078863-b7e4dec0-8d51-4c66-a23c-981ad564d659.png)

There's a large piece of time where we just see the splash screen. The reason is on a debug build the app tries to connect to the metro server - from my observations the app waits for a connection for about 5-6 sec then continues to load the bundled bytecode
- On one of my previous shots you can see the red needle around the 3.2-4s mark (after a 2sec download time for js)
- On the one here it's on the 6.4s mark (with no js download)

Something else we can see is where the `runJsBundle` takes place on the timeline
Other than that the rest of the JS timings seem to be the same

To run a debug build with precompiled JS bytecode 

1. Update `app/build.gradle` with `bundleInDebug: true` 

```gradle
project.ext.react = [
    enableHermes: true,  // clean and rebuild if changing
    bundleInDebug: true,
]
```

2. Run the app with `npm run android` but stop the metro dev server

**Comment by kidroca** [comment]
> I'd definitely be curious to know if there are some large libraries or things doing a lot of work that could be lazy loaded or deferred.

This tool can help with tracking any large JS bundles, and it can be run with `npx react-native-bundle-visualizer` 
https://www.npmjs.com/package/react-native-bundle-visualizer

I've run it on Expensify and here's what I've found out 
- total size 3.66MB (though it's not very accurate as it's actually 4.05MB)
- external libraries (from `node_modules`) - 2.7MB (61% of the source)
- `src` - 813kB - 18% of the source
- `react-native` - 586kB 13%
- `react-native-reanimated` - 321kB 7%
- `react-native-svg` - 204kB 4.5%
- `@formatJs` - 199kB 4.4%
- `@react-navigation` - 185kB 4.1%

I tool like this usually helps with tracking packages that are added 2x. or not tree shaken properly
- you can see expensify-common being added 2x 
- there's also underscore 3x (expensify-common, onyx, and the app seem to use different versions of it)
- for some reason the whole lodash is loaded (100kB) even though we only use and import specific methods from it 

To track heavy work we can use the Hermes profiling flow, or capture more metrics with `react-native-performance`

**Comment by marcaaron** [comment]
We're still working on this one Melvin. Going to make this a weekly since this is a rolling issue that will likely have more than one solution.

**Comment by MelvinBot** [comment]
@bfitzexpensify, @Luke9389 it looks like no one is assigned to work on this job.
Please double the price or add a comment stating why the job isn't being doubled.

**Comment by kidroca** [comment]
Nothing new to report here. I'm focusing on Onyx image/file handling this week

**Comment by bfitzexpensify** [comment]
Heading ooo so reassigning `External` label. This is a weekly that is going to be a rolling issue, likely with more than one solution — so if/when we hire contributors, we'll want to keep the posting open for more proposals.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @NicMendonca (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by NicMendonca** [comment]
Anyone got an update for this one? 👋

**Comment by marcaaron** [comment]
Not really AFAIK we are still waiting for specific proposals.

**Comment by marcaaron** [comment]
Let's increase the price?

**Comment by NicMendonca** [comment]
Doubled the price. Job posting[ here. ](https://www.upwork.com/jobs/~012f54c9ad750b8b1a)

**Comment by Luke9389** [comment]
N6 freeze Melvin

**Comment by NicMendonca** [comment]
@kidroca started a discussion about this issue[ here](https://expensify.slack.com/archives/C01GTK53T8Q/p1632974138285100).

**Comment by Luke9389** [comment]
Not much action on this of late.

**Comment by Luke9389** [comment]
Not much action on this of late. Should we close this?

**Comment by marcaaron** [comment]
Double the price maybe? There are probably still performance issues when starting up Android and we should definitely follow up on any ideas to make things faster. I will also say that I don't have much time to dedicate to this right now - so maybe only keep it open if others are able to give it attention and help test out proposals. Good question to ask in Slack probably.

**Comment by roryabraham** [comment]
Has anyone explored the possibility of introducing some [code splitting](https://reactjs.org/docs/code-splitting.html) to our application as a method for improving TTI?

**Comment by kidroca** [comment]
> Has anyone explored the possibility of introducing some [code splitting](https://reactjs.org/docs/code-splitting.html) to our application as a method for improving TTI?

That particular guide would not work in `react-native` as there is no dynamic `import` 

Something related Marc pointed sometime ago - our js bundle on mobile apps is compiled to bytecode which loads very fast: https://reactnative.dev/docs/next/hermes#confirming-hermes-is-in-use
> This will compile JavaScript to bytecode during build time which will improve your app's startup speed on device.

You can verify this by logging out, quitting and restarting the app 
Note how fast the login screen loads - our whole JS code is read here 
Now login, quit and restart the app
The app starts slower - but it's not because of our code being delayed 

React Navigation is doing the optimization of only loading the necessary screens and not executing the rest of the code
When it only loads the Login Screen everything runs fast
But after we authenticate a lot more code is triggered to run - running the code is slow, so even if we had some code splitting it might help a bit to load that code faster, but ultimately it's not the loading that is slow, it's the actual running

It's possible to do lazy script loading in react-native by using `require` calls, but this might be more harm than good - `require` calls block execution

We can still do some lazy loading by organizing modules in such way that their code runs only after `init` 
E.g. instead of writing `Onyx.connect` at the top level like this 

```js
let myStoragaValue;
Onyx.connect({
  key: myKey,
  callback: val => myStorageValue = val,
})
```

We do something like 

```js
let myStoragaValue;

function init() {
  Onyx.connect({
    key: myKey,
    callback: val => myStorageValue = val,
  })
}

export default init;
```

It's not clear how much performance (if any) we'll gain by going for this though
- We already make `Onyx.init` resolve the connections after it's ready. So even though there are many top level connect calls - the all are waiting and called back after `Onyx.init`

**Comment by NicMendonca** [comment]
👋 any update here?

**Comment by Luke9389** [comment]
The main update is that we've outsourced performance to a group that will be working with us starting in April.

**Comment by NicMendonca** [comment]
Hi everyone! What shall we do this?

**Comment by Luke9389** [comment]
We can keep it open for now. We'll end up assigning it to our outsourced performance group when they start.

**Comment by Luke9389** [comment]
Still waiting for them to start. (Should be June)

**Comment by Luke9389** [comment]
Callstack has sent an updated start date and team composition. Should be August.

**Comment by Luke9389** [comment]
Callstack situation keeps getting delayed. I'll post here when we have a firm start date

**Comment by Luke9389** [comment]
Looks like we're dropping callstack. I'm going to unassign.

**Comment by marcaaron** [comment]
Let's maybe see if anyone wants to lead this up internally. There's a PR out that improved the TTI on Android by like 800ms atm. So unsure how much faster we are looking to make things beyond that.

**Comment by Luke9389** [comment]
I know @AndrewGable has been talking with callstack so perhaps he'd be interested in taking this.

**Comment by JmillsExpensify** [comment]
I also just highlighted this issue in Slack as a performance issue in need or a volunteeer.

**Comment by trjExpensify** [comment]
@marcaaron - which PR is that? https://github.com/Expensify/App/issues/4656#issuecomment-1247373020

**Comment by marcaaron** [comment]
This one here to lazy load navigation screens -> https://github.com/Expensify/App/pull/10986

**Comment by marcaaron** [comment]
Noticing that this issue is eroding a bit and not getting much love. Maybe it's not clear what we are trying to achieve?

**Ideas for how to take action on this...**

- It would be helpful if someone could work on getting some baseline measurements together by building release versions of Android and iOS apps
- Update the description with where we are at on those platforms
- Land on some kind of realistic target together for how fast the app should open

**Comment by roryabraham** [comment]
Just throwing this out there ... we have [this issue](https://github.com/Expensify/App/issues/8503) to enable the new architecture. Part of that architecture, maybe less-talked-about than the new [Fabric renderer](https://reactnative.dev/architecture/fabric-renderer), is called [TurboModules](https://reactnative.dev/docs/next/the-new-architecture/pillars-turbomodules). With the traditional React Native architecture all native modules had to be loaded up-front before the app would load and the JavaScript would start running. With TurboModules, native modules are lazy-loaded – that _should_ bring forth some "free" reductions in TTI.

**Comment by mountiny** [comment]
@roryabraham I feel like this issue live kind of in vacuum. Should we either make this a tracking issue for PRs/issues which are aimed to improve the app loading time or I think we might want to close this one out as Margelo already addressed some boot up changes and we might want to go in a direction of creating a specific issues for specific problems (similar as you have linked above - enable fabric to speed things up)

This issue is from august 2021 so I feel like it is a bit out of context at the moment. I think we could close this out, if you agree, feel free to close this one out.

**Comment by JmillsExpensify** [comment]
@mountiny Agreed. I'll go ahead and link this one to the existing [tracking issue](https://github.com/Expensify/App/issues/10962).

**Comment by marcaaron** [comment]
Idea for clearer deliverable -> 

>Land on some kind of realistic target together for how fast the app should open

Part of the problem with this issue is it's not clear how much faster things should be and an audit of where things are at would be a good place to start.

**Comment by roryabraham** [comment]
I agree @marcaaron, and think that [this issue](https://github.com/Expensify/App/issues/11997) could be a great starting place.

**Comment by kidroca** [comment]
I think we talked about multiget on a few occasions but we never really gave it a chance and it might boost performance
- https://github.com/Expensify/App/issues/4433

A brief summary of the idea is
1. We pretty much read all (or almost all) Onyx finite ("small") keys in memory at launch (or soon after)
2. Instead of creating parallel read calls for each such key - we can request them with one `multiGet` operation
